Homework 5

DaniilVyatkin opened the pull request:
Homework 5. Implemented RemoteControl and commands for it.

DaniilVyatkin committed:
Homework 1 on SmartHome

DaniilVyatkin committed:
Homework 2
Unfortunately, I was unable to correctly implement turning off all the lights on closing the hall door.

DaniilVyatkin committed:
Implemented 2 adapters: from CCSensorEvent to SensorEvent and vice versa.

DaniilVyatkin committed:
Implemented eventHandlerToCCEventHandlerAdapter and added initialization using @Beans from MyConfiguration in main() in Application.

DaniilVyatkin committed:
Implemented RemoteControl and commands for it.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Application.java:
```diff
@@ -1,80 +1,68 @@
+                Arrays.asList(new DoorEventHandler(), new LightEventHandler(), new HallDoorEventHandler()));
+
+        // начинаем цикл обработки событий
+        while (true) {
```
См. https://github.com/iagoshkov/smart-home-2020/blob/master/STUDENT_GUIDE.md#перенесите-логику-цикла-обработки-событий-в-отдельный-класс с первого дз не исправлено.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/ProgrammableRemoteControl.java:
```diff
@@ -0,0 +1,19 @@
+import java.util.HashMap;
+
+public class ProgrammableRemoteControl implements RemoteControl {
+    HashMap<String, CommandRemoteControl> buttonMap = new HashMap<String, CommandRemoteControl>();
```
`Map<String, CommandRemoteControl> buttonMap` без деталей реализации.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/ProgrammableRemoteControl.java:
```diff
@@ -0,0 +1,19 @@
+    HashMap<String, CommandRemoteControl> buttonMap = new HashMap<String, CommandRemoteControl>();
+    @Override
+    public void onButtonPressed(String buttonCode) {
+        buttonMap.get(buttonCode).execute();
```
Проверка на присутствие buttonCode.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/ProgrammableRemoteControl.java:
```diff
@@ -0,0 +1,19 @@
+public class ProgrammableRemoteControl implements RemoteControl {
+    HashMap<String, CommandRemoteControl> buttonMap = new HashMap<String, CommandRemoteControl>();
+    @Override
+    public void onButtonPressed(String buttonCode) {
```
Оригинальная сигнатура интерфейса по заданию была `void onButtonPressed(String buttonCode, String rcId);`, зачем вы ее изменили?

DaniilVyatkin replied:
На лекции (28 октября, примерно на 1:24:50) Илья сказал, что нужно убрать отсюда rcID.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/ProgrammableRemoteControl.java:
```diff
@@ -0,0 +1,19 @@
+        buttonMap.get(buttonCode).execute();
+    }
+
+    public void setButton(String buttonCode, CommandRemoteControl commandRemoteControl) {
```
Это не надо делать, просто передайте готовую мапу через конструктор.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Application.java:
```diff
@@ -1,80 +1,68 @@
+        Application application = new Application(homeLoader, eventGenerator);
+
+        // Добавим пульт управления
+        ProgrammableRemoteControl remoteControl = new ProgrammableRemoteControl();
```
По 4 и 5 дз мы используем спринг, добавьте конфигурацию, см. пример main в презентации к 4 дз.

DaniilVyatkin committed:
Homework 5 corrections
1. Вынес цикл обработки событий из Applicatoin в отдельный класс (как в 1ом ДЗ)
2. Исправил HashMap на Map для buttonMap в ProgrammableRemoteControl
3. Добавил проверку на присутствие buttonCode в ProgrammableRemoteControl
4. Заменил setButton на передачу мапы с кнопками через конструктор в ProgrammableRemoteControl
5. Добавил конфигурацию RemoteControlRegistry в MyConfiguration

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/RCcommands/ActivateAlarmCommandRC.java:
```diff
@@ -0,0 +1,7 @@
+
+public class ActivateAlarmCommandRC implements CommandRemoteControl {
+    @Override
+    public void execute() {
```
Где сигнализация из 3 дз? Где команда по ее активации?

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Application.java:
```diff
@@ -1,80 +1,69 @@
+        Application application = new Application(homeLoader, eventGenerator, eventHandlingLoop);
+
+        // Настроим кнопки для пульта
+        HashMap<String, CommandRemoteControl> buttonMap = new HashMap<String, CommandRemoteControl>(Map.ofEntries(
```
Это все по 4-5 дз необходимо перенести в spring конфигурацию.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/Application.java:
```diff
@@ -1,80 +1,69 @@
+
+        // Я правильно понял, что теперь мы вместо нижележащего кода (4 строчки)
+        // используем SensorEventsManager из com.coolcompany.smarthome.events,
+        // который делает примерно все то же самое, но без загрузки дома?
```
Да, см. как должен выглядеть main в слайде к 4 дз.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MyConfiguration.java:
```diff
@@ -0,0 +1,35 @@
+public class MyConfiguration {
+    @Bean
+    SensorEventsManager sensorEventsManager() {
+        SensorEventsManager sensorEventsManager = new SensorEventsManager();
```
Это все переделать по замечаниям к 4 дз.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/MyConfiguration.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Bean
+    RemoteControlRegistry remoteControlRegistry() {
+        return new RemoteControlRegistry();
```
Тут еще надо регистрировать RemoteControl с вашей реализацией.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/RCcommands/CloseHallDoorCommandRC.java:
```diff
@@ -0,0 +1,29 @@
+                Collection<Door> doors = room.getDoors();
+                for (Door door : doors) {
+                    CloseDoorAction closeDoorAction = new CloseDoorAction();
+                    closeDoorAction.setId(door.getId());
```
Вместо сеттера setId можно передать door.getId() в конструктор CloseDoorAction.

a-ivanov commented on src/main/java/ru/sbt/mipt/oop/RCcommands/CloseHallDoorCommandRC.java:
```diff
@@ -0,0 +1,29 @@
+
+    @Override
+    public void execute() {
+        Collection<Room> rooms = smartHome.getRooms();
```
Метода **getRooms** в доме не должно быть со 2 дз. Он заменен на применение Action к внутренним объектам дома. Переделайте.

DaniilVyatkin commented:
Исправления по 5 дз я добавил в ветку Homework-4, потому что там все актуальные изменения по остальным дз. Я не уверен, что можно аккуратно заменить весь код в ветке Homework-5 на код из ветки Homework-4 и при этом не сломать историю и комметарии.
